### ISSUE TITLE
RCP14/OSA : repo build still broken with optional components (swift)

### ISSUE THREAD

**Issue by BjoernT** [rcbops/rpc-openstack#2027]
###### Please provide the output to 'git describe --all' for  all deployments where you have confirmed it occurs.

newton-14.0 branch
OSA: 14.1.0-16-g9a14560

##### What type of issue is this? Please tick the appropriate box.
- [x] bug - existing functionality that does not work for you as documented
- [ ] feature - new functionality that you would like to see added

##### What is the expected behavior?

No build errors

##### What is actually happening?

Selective build process still seem to have issues per
https://bugs.launchpad.net/openstack-ansible/+bug/1671958

##### What were you attempting to do when you encountered this error (give as much detail as possible)? 

Fresh install newton

##### How can we reproduce this issue?

N/A

##### How can we verify your issue is addressed?

No errors

##### If you are using or know of a workaround please describe the necessary steps?

N/A

##### What is your evaluation on the importance of this issue, and why?

Prio 2

----------------------------------------------------------------------------------------------------------------------------
###### For each pull request, update the branch line to reference it.
- [ ] master
- [ ] newton-14.0

##### Upgrade impact. Consider the impact on moving between major, minor and patch releases.
- [ ] This change does not impact upgrades
- [ ] This change addresses a potential upgrade impact
- [ ] The upgrade impact is being tracked by another issue.

The issue tracking the upgrade impact is:

##### Testing impact
- [ ] An issue/pull-request has been create or exists to track the testing gap, and is referenced below
- [ ] This change includes additional testing
- [ ] This change is tested by the existing setup

Additional testing links:

##### Release note
- [ ] This change includes a release note.
- [ ] This change does not require a release note.

##### Does your change require documentation?
- [ ] Yes, and I submitted an issue to [Docs issues](https://github.com/rackerlabs/docs-rpc/issues "Docs issues")
- [ ] Yes, and I contributed a documentation patch to [docs-rpc](https://github.com/rackerlabs/docs-rpc "docs-rpc")
- [ ] No documentation required for this change

Additional documentation links (please add links if you have ticked yes):

**Comment by claco** [comment]
```
What were you attempting to do when you encountered this error (give as much detail as possible)?

Fresh install newton
```

I assume you mean a fresh install of Newton rc1 _without_ Swift enabled?

Can you provide the config for that for replication? What this a multi node, or an AIO?

I _think_ @Frank-ZhangXin and @prometheanfire ran into something like this Friday working on something else, with @d34dh0r53 troubleshooting with them.

**Comment by Frank-ZhangXin** [comment]
@BjoernT Hey man, could you check if the environment variable 'DEPLOY_SWIFT' equals to 'yes'? Our issue is caused rebooting then previous setting is lost.

**Comment by claco** [comment]
@BjoernT IS the use case for this that the install is using cloud files as the backend store?

**Comment by claco** [comment]
@BjoernT @alextricity25 @cloader89 Likely related:  https://github.com/rcbops/rpc-openstack/issues/1659

**Comment by BjoernT** [comment]
Yes I was doing a fresh install without swift on the newton-14.0 branch.
I am using NFS as glance backend.

And no I'm not using deploy.sh so DEPLOY_SWIFT does not apply since it's only used for bootstrap-aio.sh.

As workaround I did set

`repo_build_wheel_selective: false`

**Comment by claco** [comment]
I thought we were all (dev/qe/deployments) were all going to use deploy.sh all the time?

**Comment by BjoernT** [comment]
@claco That doesn't matter I used it at first, the problem is inside the repo_build role, so how ever you call it is secondary

**Comment by b3rnard0** [comment]
General consensus is for us to do the a documented workaround as @BjoernT noted and to verify that this workaround is working. A specific fix will be targeted for r14.0.1 revision release.

@claco @BjoernT thoughts?

**Comment by claco** [comment]
@BjoernT @b3rnard0 @alextricity25 @cloader89 I guess it depends on the LOE we think it might take in the MaaS role to avoid using repo_build_wheel_selective. Also, now many Greenfields to we anticipate installing w/o Ceph and w/o Maas, just using NFS/Cloud Files as the backend?

**Comment by odyssey4me** [comment]
The reason that this is happening is because the selective build makes a decision about packages to build based on host group membership, and swift has no host members. The reason that the swift package is trying to build in the first place is because it's listed in maas_pip_packages for the rpc_maas role. A good solution here would be to only include that package is swift is installed.

**Comment by claco** [comment]
Yup. I think that's basically the bug @alextricity25 filed that we just didn't fix as part of our Newton work. Also, testing gap, so it didn't crop up again on the non human side of testing.

**Comment by cloader89** [comment]
@claco the reason no one saw the error until @BjoernT pointed out, our deploy.sh does not remove the swift.yml file from `/etc/openstack_deploy/conf.d/` so swift is always being configured on AIO's, thus always passing until manually setting up the environment.

Accoding to mitaka13.1:
https://github.com/rcbops/rpc-openstack/blob/mitaka-13.1/scripts/deploy.sh#L131-L136

Newton:
https://github.com/rcbops/rpc-openstack/blob/newton-14.0/scripts/deploy.sh

Might have a separate issue according to above.

Issue is confirmed.

Ran the workaround.  I had to add the variable referenced inside `repo-build.yml`. This provides a successful repo-build.

**Comment by odyssey4me** [comment]
I have proposed a patch to OSA to resolve this: https://review.openstack.org/445922

The suggested workaround, if you need it because you don't want to wait for a SHA bump, is to add ``repo_build_git_selective: no`` to ``rpcd/etc/openstack_deploy/user_osa_variables_defaults.yml`` so that it copies down into the user config of the deployment. That is the most minimal change to implement which will not affect the repo build time as much as turning off the selective wheel build.

**Comment by cloader89** [comment]
Awesome @odyssey4me thanks. Should we code the workaround, or hold off? @b3rnard0 @claco

**Comment by claco** [comment]
@cloader89 We already seemed to have reached an agreement on 14.0.1 and not 14.0.0rc for a fix time.  https://github.com/rcbops/rpc-openstack/issues/2027#issuecomment-286128196

Seems like we have time to wait for it to propagate down to Newton upstream and bump the sha. Of course, we still need to ensure it's documented.

**Comment by b3rnard0** [comment]
@robb-romans and/or @erikmwilson This is going to go into 14.0.1 as a fix but we need a workaround release note for 14.0.0. Please reach out to @cloader89 for questions.

**Comment by odyssey4me** [comment]
FYI the backports have been implemented: https://review.openstack.org/#/q/7331b821b4e2894a62bd22e7e47e87e04b7d7f71,n,z

**Comment by asettle** [comment]
Doc release note: https://github.com/rackerlabs/docs-rpc/pull/989

**Comment by alextricity25** [comment]
@odyssey4me It sounds like this can be moved over to the "Pending SHA Update" column. Is that correct?

**Comment by odyssey4me** [comment]
@alextricity25 Yep, unless it needs to move to some sort of docs column. I'm not altogether familiar with the workflow being applied here.

**Comment by alextricity25** [comment]
@odyssey4me Great! Thanks for making the fix for this!

**Comment by alextricity25** [comment]
I believe this will need a good ol' role sha bump upstream before we can do the sha bump here.

**Comment by alextricity25** [comment]
SHA bump to include this fix in master is here:
https://github.com/rcbops/rpc-openstack/pull/2077

I'll update this card once we get a SHA bump for newton-14.0 going.

**Comment by alextricity25** [comment]
The SHA bump has gone through on master. Now we need a SHA bump on newton-14.0 for this fix to be included in RPCO's newton releases.

**Comment by alextricity25** [comment]
UPDATE:
I beilieve this is still pending a SHA update on the newton-14.0 branch.

**Comment by cloudnull** [comment]
@alextricity25  is this something we need to take a look at within OSA?

**Comment by alextricity25** [comment]
This is resolved in OSA. However, the SHA in our newton-14.0 branch doesn't include the change.
Once the OSA SHA is bumped in the newton-14.0 branch, then this can be moved out of "status-pending-sha-updated" and be closed. Thanks @cloudnull!

**Comment by alextricity25** [comment]
This is now included in all the appropriate branches. Thank you all!
